1단계 - 엔티티 매핑

SHINMJ opened the pull request:
JPA 미션 1단계 다음과 같이 구현했습니다.
- 엔티티 클래스 매핑
- DataJpaTest 활용한 테스트 케이스 작성
- Jpa Auditing 기능 추가 (createAt, updateAt)

SHINMJ committed:
docs: readme.md 생성

- README.md 추가
- 기능 구현 목록 작성

SHINMJ committed:
feat, refactor: 엔티티매핑

- application.properties -> yml 변경
- sql 확인 property, diarect Mysql property 추가
- JPA Auditing 클래스 BaseTime.java 추가
- 도메인 클래스 entity 매핑

SHINMJ committed:
docs: README.md update

- 기능 구현 목록 업데이트

SHINMJ committed:
docs,feat,test: README.md update, 엔티티 클래스 매핑,test case 추가

- 기능 구현 목록 update
- 엔티티 클래스 매핑
- user repository 테스트 코드 추가

SHINMJ committed:
test: Answer test case add

- 답변 엔티티 테스트 케이스 추가
- 답변 repository 테스트 케이스 추가

SHINMJ committed:
test: 질문,답변, 삭제히스토리 테스트 케이스 작성

na-mul commented on README.md:
```diff
@@ -0,0 +1,20 @@
+# QNA JPA
+
+## 기능 구현 목록
```
요구사항에 굳이 명시되어 있지 않았음에도 기능목록 정리 👍입니다.

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,13 +3,27 @@
     private ContentType contentType;
     private Long contentId;
     private Long deletedById;
     private LocalDateTime createDate = LocalDateTime.now();
```
기본값을 잘 설정해주셨지만 결국 생성자에서 `createDate`를 받고 있기 때문에, 
생성자로 넘어오는 값으로 설정될 것으로 보입니다.

- `LocalDateTime.now()`를 강제하고 싶다면 생성자에서 `createDate`를 제거해보는 것은 어떨까요? 
혹은 그 반대도 가능합니다.

비록 과제의 기본코드에서 처음부터 그렇게 되어 있었다 할지라도 , 적극적으로 코드 수정을 해주셔도 괜찮아요!
옳다고 생각하시는 방향으로 기본 코드를 마구마구 손대주시길 부탁드립니다 🙇‍♂️🙏

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,13 +3,27 @@
     private Long deletedById;
     private LocalDateTime createDate = LocalDateTime.now();
 
+    public DeleteHistory() {
+    }
+
```
`Setter`가 없기 때문에 아마 이 호출자를 사용할 일은 없을 것으로 보입니다.
그렇다면 접근제어자를 `protected` 등으로 바꿔보면 어떨까요?

- 다른 엔티티들도 기본생성자가 public으로 열려있지는 않은지도 점검 부탁드릴께요.

na-mul commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,12 +1,33 @@
+    @Column(length = 100, nullable = false)
     private String title;
+
+    @Column(columnDefinition = "LONGTEXT")
```
`contents`가  `clob` 이어야 된다는 요구사항을 잘 반영해주셨군요!

na-mul commented on src/main/java/qna/domain/common/BaseTime.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.data.annotation.LastModifiedDate;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@MappedSuperclass
+@EntityListeners({AuditingEntityListener.class})
+public class BaseTime {
+	@CreatedDate
+	private LocalDateTime createdAt;
+	@LastModifiedDate
```
Spring Data JPA를 사용하셔서 auditing을 잘 적용해주셨네요!

na-mul commented on src/test/java/qna/domain/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,65 @@
+		});
+	}
+
+	@Test
+	@DisplayName("질문 찾기 성공")
+	public void findQuestionByIdSuccess() {
+		questionRepository.save(QuestionTest.Q1);
+		Optional<Question> optionalQuestion = questionRepository.findByIdAndDeletedFalse(QuestionTest.Q1.getId());
+
+		assertAll(() -> {
+			assertThat(optionalQuestion.isPresent()).isTrue();
+			Question question = optionalQuestion.get();
+			assertThat(question.getTitle()).isEqualTo(QuestionTest.Q1.getTitle());
+		});
+
+	}
```
단독으로 수행시에는 문제 없지만, 클래스 내의 전체 테스트를 한번에 돌리게 되면 실패가 발생하는 테스트로 보입니다.
- 원인을 찾아주시길 부탁드릴께요.

그리고 각 테스트는 타 테스트에 의존적이면 안된다는 원칙을 잘 기억해주시면 감사하겠습니다! 🙇‍♂️

na-mul commented on src/test/resources/application.yml:
```diff
@@ -0,0 +1,16 @@
+spring:
+  profiles:
+    active: test
+  jpa:
+    hibernate:
+      ddl-auto: create-drop
+    properties:
+      hibernate:
+        format_sql: true
+logging:
+  level:
+    org:
+      hibernate:
+        type:
+          descriptor:
+            sql: trace
```
test 전용 yml 파일을 잘 설정해주셨군요! 덕분에 테스트가 문제없이 잘 수행됩니다 👍

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,13 +3,27 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
     private ContentType contentType;
```
`enum`의 경우 `@Enumerated(EnumType.STRING)` 어노테이션을 꼭 활용해주는 편이 좋습니다!
그렇지 않으면 다음과 같이 데이터가 들어가게 되요..

![image](https://user-images.githubusercontent.com/58849938/140763837-d6712c2a-bdd9-4aa5-bd0d-198469dc770f.png)

위 캡쳐처럼 `contentType`이 enum 순서를 나타내는 숫자로 세팅되는 것이 default입니다.

수정을 부탁드릴께요.

na-mul reviewed (changes requested):
안녕하세요 명진님! JPA 미션을 함께하게 될 노동환입니다.
제출해주신 코드 잘 보았어요. ^^
간단한 코멘트를 달아두었으니 한번 확인하시고 재차 request 부탁드리겠습니다! 🙇‍♂️

SHINMJ committed:
refactor: 리뷰 참고 리팩토링

- enum type 매핑 시 enumerated 선언
- 엔티티 생성자에대한  접근 제어자 변경
- 불필요한 파라미터 제거
- 실패한 테스트 케이스 수정

SHINMJ committed:
style: formatting

- line, indent 등 포맷팅 수정

na-mul commented on src/test/java/qna/domain/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,65 @@
+	@Autowired
+	private QuestionRepository questionRepository;
+
+	@AfterEach
+	public void tearDown() {
+		questionRepository.deleteAll();
+	}
```
`deleteAll()`을 통해 매 테스트마다 data는 지워져도, 생성되는 id까지 돌아가지는 않습니다. 
앞선 테스트에서 id가 1,2인 채 생성되었으면, deleteAll()을 해도 그 다음의 테스트에서는 엔티티의 id가 3부터 생성되게 됩니다.

따라서 QuestionTest.Q1, Q2등으로 공통된 객체를 여러 테스트에서 동시에 사용할 때는 주의 깊게 사용하여야 합니다.
이 점은 참고만 부탁드립니다~🙇‍♂️

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,110 @@
+
+@Entity
+public class Answer extends BaseTime {
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private Long id;
+	private Long writerId;
+	private Long questionId;
```
들여쓰기를 일괄 변경해주신 이유가 궁금합니다!
그 전에더 4 space 로 되어있어서 크게 문제될 것은 없어보였거든요!

na-mul reviewed (approved):
안녕하세요 명진님. 코멘트 드린 내용 잘 반영해주셔서 감사합니다 😄
이번 기회에 더욱 JPA에 능숙해지셔서, 명진님 하시는 일에 큰 힘이 되었으면 하는 바람입니다.
다음 단계도 기대되네요!
